### PR TITLE
Collect APM metrics for failure stores

### PR DESCRIPTION
**Pull request by nielsbauman**
This PR adds APM metrics for failure stores. See the JavaDoc comments in `FailureStoreMetrics.java` for a detailed explanation on the individual metrics.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 1]
```diff
@@ -879,7 +884,9 @@ private void executePipelines(
         ingestDocument.resetReroute();
         final String originalIndex = indexRequest.indices()[0];
         final Consumer<Exception> exceptionHandler = (Exception e) -> {
+            FailureStoreMetrics.incrementForIndex(failureStoreMetrics.failureCounter(), originalIndex);
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java:887]: how about
`failureStoreMetrics.incrementFailureOn(originalIndex)` / `failureStoreMetrics.incrementRedirectOn(originalIndex) `

**Review by pgomulka — APPROVED** [review]
it looks good - i left 1 suggestion
I can see that it requires quite a lot of wiring to have the metric reused in both IngestService,BulkOperation. I don't think there is an easy way to workaround this now - so I think it is ok to have the `FailureStoreMetrics`

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 1 above)
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:887]: Out of curiosity, what makes you prefer your suggestion over my current implementation? I like that I only have to have one method, but my method invocation line is a bit longer in my version.

Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:887]: it looks like you have added a static utility method on a class to operate only on its instances. It would feel more natural to have this as an instance method.

>  I like that I only have to have one method, but my method invocation line is a bit longer in my version.

you have one method, but you can only use it with 2 parameters (failureCounter/redirectionCounter). Any other instance of counter would not be valid. I think it is better to make it explicit with 2 separate method names.

Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:887]: Ah okay that makes sense, thanks for the explanation. Will update :+1:

File: server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    public static final String METRIC_REDIRECT_TOTAL = "es.ingest.failure_redirects.total";
+
+    public FailureStoreMetrics(MeterRegistry meterRegistry) {
+        this(
+            meterRegistry.registerLongCounter(METRIC_FAILURE_TOTAL, "ingest failure counter", "unit"),
+            meterRegistry.registerLongCounter(METRIC_REDIRECT_TOTAL, "ingest failure redirect counter", "unit")
+        );
+    }
+
+    public static void incrementForIndex(LongCounter counter, String index) {
```
Comment (jbaiera) [inline server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:33]: I wonder if perhaps we should capture if the failure is an ingest node failure or a shard level failure. Can we think of any other attributes that might be useful to capture?

**Review by jbaiera — COMMENTED** [review]
Adding some initial thoughts about the metrics.

File: server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.elasticsearch.telemetry.metric.LongCounter;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+
+import java.util.Map;
+
+/**
+ * A record APM metrics that concern failure stores.
+ */
+public record FailureStoreMetrics(LongCounter failureCounter, LongCounter redirectCounter) {
```
[the remark this answers is not in the document]
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:19]: I saw this approach in [RepositoriesMetrics.java](https://github.com/nielsbauman/elasticsearch/blob/08b687811baac42d7d105f0fa9aca98882728534/server/src/main/java/org/elasticsearch/repositories/RepositoriesMetrics.java#L15-L15) and I thought it looked clean. Would you prefer a regular class?

File: server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java (hunk 2 above)
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:33]: I also think this is a good metric to have.

If we track metrics per index, should we also track per pipeline?

Also, should we track per type of error? Initially, I would think that this would vary too much, but if we split the metrics by index name, then I guess the error type is not that crazy either.

**Review by gmarouli — COMMENTED** [review]
Hm, it's not clear to me why we count failures of non failure store indices and data streams in the failure store metrics.

If this is indeed what we want, I would prefer to rename this to something that better captures what we track here.

Does this make sense to you, or am I missing something?

**Comment by nielsbauman** [comment]
@gmarouli I could see value in users getting insights into which data streams see a lot of failures and could thus benefit from a failure store. I also don't immediately see value in tracking failure counts for non-data stream indices, so it probably makes sense to just track the failures for data streams only. Maybe @jbaiera can elaborate on this?

File: server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java (hunk 2 above)
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:33]: I'm inclined to agree that it makes sense to add all these attributes. My only question/concern is that I don't know how these metrics are stored exactly, and thus what the impact would be of adding attributes (mainly in terms of memory consumption I guess). Is either of you familiar with the inner workings of APM? If not, I can do some investigation.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 4]
```diff
@@ -879,7 +884,11 @@ private void executePipelines(
         ingestDocument.resetReroute();
         final String originalIndex = indexRequest.indices()[0];
         final Consumer<Exception> exceptionHandler = (Exception e) -> {
+            if (originalIndex != null) {
+                FailureStoreMetrics.incrementForIndex(failureStoreMetrics.failureCounter(), originalIndex);
+            }
```
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java:889]: Apparently, during a bulk upsert ([example](https://github.com/nielsbauman/elasticsearch/blob/46770e7af2c0146ac411376d566d5d0d1791b770/modules/ingest-common/src/yamlRestTest/resources/rest-api-spec/test/ingest/60_fail.yml#L105-L105)) we don't have an index name in the index request.

File: server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbaiera) [reply server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:19]: It looks like RepositoriesMetrics is only ever used to hold the counters. Since we're expecting people to increment the counters in the FailureStoreMetrics via the methods on it, we probably shouldn't expose the counters via getters. I think a regular class makes more sense here if we make the incrementing methods non-static.

File: server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java (hunk 2 above)
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:33]: Sorry, I would not expect that counters would be that heavy, but we should understand how these metrics could grow, ballpark figures I mean.

Comment (jbaiera) [reply server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:33]: > Full disclosure, part of the reason I'm having second thoughts is that it's slightly inconvenient to get the error type and pipeline, as we have to perform some checks to able to obtain them

I'm alright with leaving ingest node specific stats off for now. If we find that we need that data we can try to add them to the attributes in a follow up.

> I'm not sure all these attributes are actually valuable. Wouldn't these attributes overlap with/overshadow a part of the purpose of the failure store? In other words, what do both of you see as the value of adding these attributes?

I think having a way to track these failure stats when managing multiple clusters would be beneficial. Failure stores could hold sensitive information in them, so operators and maintainers may not always have access or permission to view the failure store. Collecting metadata from multiple clusters about failure rates while avoiding identifying information so that it can be analyzed in a central place is likely worth the double counting.

> I wonder if perhaps we should capture if the failure is an ingest node failure or a shard level failure. Can we think of any other attributes that might be useful to capture?

I think capturing the following attributes would be beneficial:

* The data stream we are writing the data to
  * Lets us see what kinds of integrations are experiencing problems
* The kind of failure we're experiencing (ingest node|shard level)
* The exception type that we've encountered
  * This helps us identify the failure without exposing any sensitive information about the failed documents
* Whether or not the data stream we are writing the data to has failure stores enabled or not
  * Failure stores will likely never be on for every data stream by default and users will have the option of disabling them on integrations. This helps us identify where any gaps in our coverage may be, and which integrations may be dropping failed data on the floor.

In the future I could see the following be helpful to trace:
* Ingest pipeline name and the processor type+tag to help identify specific problems
* If our errors ever end up with a unique id for their cause we could capture that (like 
* Some day we could try and capture which field is causing the problem with mappings?

Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:33]: I think what you're saying definitely makes sense to me. I do have some follow-up thoughts/questions:
- Do we still see value in tracking the number of failures that got redirected? 
- Does it make sense to _only_ track the number of failures for data streams that do not have failure stores enabled? Neither in the scenario you sketched nor in any I can think of right away do I see great value in tracking those numbers for data streams that do not have failure stores enabled. 
- As @gmarouli pointed out, it probably doesn't make sense to track these numbers for non-data stream failures at all, correct?

**Comment by felixbarny** [comment]
I think it would be important to capture the following information:

* Counter of rejected docs per index
* Counter of documents that have landed in the failure store
  * Information about the category of the failure (like the exception type), and whether it has happened in the pipeline vs a mapping issue. Maybe that's the same.
* Counter of documents that were indexed but are degraded, in the sense that it has an `_ignored` field
* Counter of total documents that were requested to be indexed per index, to allow calculating the percentage of failed documents

One doubt that I have is whether the index name may be a high cardinality attribute so that we have to track lots of different metrics in memory if there are a lot of indices. But since we have per-index metadata in memory anyway, it may not be a big issue. Maybe we need to take care that we only track these metrics on the node that holds a primary shard for an index.

Also cc @ruflin @flash1293 to provide input on which ingestion/failure store metrics are relevant.

**Comment by flash1293** [comment]
@felixbarny We discussed this topic last week with @jbaiera in the ES+Obs sync and concluded that for the time being we can count the docs by looking at index stats and grouping them in a certain way as part of a background job running in Kibana. This is how we are collecting data telemetry already, so it would require the least amount of changes as it's mostly an extension of the existing setup - it also has the additional advantage that it would work for all environments including on-prem.

A similar thing is discussed here for the `_ignored` field: https://github.com/elastic/elasticsearch/issues/108092

These two things together would cover most of the telemetry needs from Observability we know about right now. This document is laying out our plans: https://docs.google.com/document/d/1Qy8DlsJTVfi8zDvZMAkOxjvh6zsXFJRJYR_aRc7eGLQ/edit#heading=h.1qaayw9lp75l

This isn't meant to block these efforts, just as an FYI for the immediate plans on that front.

**Comment by felixbarny** [comment]
One tradeoff to be aware of is that we then can't track the number and ratio of rejected docs as they'll never make it into the system.

**Comment by jbaiera** [comment]
> Counter of rejected docs per index

@felixbarny Just want to clarify that you mean documents that have failed and their failures will **not** be collected in the failure store?

There are some cases where a failure _would_ be directed to the failure store, but due to availability issues it may be returned to the user with an error instead. Should we track those documents as well?

> Counter of documents that were indexed but are degraded, in the sense that it has an _ignored field

I don't think we'll be able to collect these stats from where we're focusing in this PR - this will have to be done closer to the mapping code since the documents are accepted without any notification to the client that they have degraded.

> One doubt that I have is whether the index name may be a high cardinality attribute so that we have to track lots of different metrics in memory if there are a lot of indices.

I checked with core infra a while back about the metrics collection and how it handles high cardinality. Their read on the situation is that the cardinality shouldn't be an issue as long as it's not part of the counter name and only part of the attributes on the counter when incrementing it.

> Maybe we need to take care that we only track these metrics on the node that holds a primary shard for an index.

Most of these stats will need to be collected when handling the bulk request on the coordinating node, and since any node can coordinate a bulk operation it's unlikely we'll be able to constrain the collection to the primary shards 😞

File: server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java (hunk 2 above)
Comment (jbaiera) [reply server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:33]: > Do we still see value in tracking the number of failures that got redirected?

I think it's still valuable to collect these for the sake of transmitting them via APM. Otherwise we'll need to source it from something like the stats, and those stats may have different retention times than those we collect from APM.

> Does it make sense to only track the number of failures for data streams that do not have failure stores enabled? Neither in the scenario you sketched nor in any I can think of right away do I see great value in tracking those numbers for data streams that do not have failure stores enabled.

Tracking failed documents for data streams without the failure store enabled would allow us to identify how much data we're rejecting outright that could have been saved into a failure store. @felixbarny do you have strong feelings about this point?

> As @gmarouli pointed out, it probably doesn't make sense to track these numbers for non-data stream failures at all, correct?

Let's keep them to data stream failures only. Let's not track requests to standalone indices. Agreed there.

**Comment by jbaiera** [comment]
> Counter of total documents that were requested to be indexed per index, to allow calculating the percentage of failed documents

Do we want to track this for all data streams regardless of if they have failure stores enabled? Collecting stats about data streams without the feature enabled should make the metric name more generic, but if we only care about totals to data streams with failure store enabled it should be a more specific metric name.

@felixbarny Just make sure we're aligned on the possible counters for this PR, here are the things that I think we could reasonable track right now along with some suggested names for the metrics. Please chime in on the metric name suggestions because they likely need to be refined still.

Counters
* Total documents sent to a data stream
  * All data streams or just those with failure store enabled?
  * Suggestion `es.data_stream.ingest.documents.total` if we want all data streams regardless of failure store enabled
* Rejected documents
  * Suggestion `es.data_stream.ingest.documents.rejected`
* Failures that could have been stored but wont because failure store is disabled (?)
  * Suggestion `es.data_stream.ingest.documents.failed` - maybe with the `failure_store_enabled` attribute set false?
* Documents sent to the failure store
  * Suggestion `es.data_stream.ingest.documents.failed`
* Documents rejected from the failure store (?)
  * Suggestion `es.data_stream.ingest.documents.failure_store.rejected`

Attributes to trace
* Data stream name (`data_stream_name`)
* Exception Type (`error_type`)
* Pipeline vs Mapping issue (`error_location` (?))

Out of scope: 
* Count of degraded documents (`_ignored` field)
  * This isn't really failure store related - this data is technically successfully indexed. We'll need to track this somewhere else.

**Comment by felixbarny** [comment]
>> Counter of rejected docs per index
>
> @felixbarny Just want to clarify that you mean documents that have failed and their failures will not be collected in the failure store?

Yes, that's what I meant. The intention behind that is that we can find out how well we're tracking against the "accept all logs" principle. When the failure store is enabled, we'd expect the ratio of rejected documents to go down significantly. But the failure store isn't enabled by default - at least in the beginning. It can also help to identify leaks where the failure store was supposed to catch a failure but didn't for some reason.

>> Counter of documents that were indexed but are degraded, in the sense that it has an _ignored field
>
> I don't think we'll be able to collect these stats from where we're focusing in this PR

Agree that this would be outside the scope of this PR. At the same time, I wanted to look at the problem more holistically in terms of tracking ingestion issues vs just tracking failure store metrics. Tracking the failure store is part of it but it's part of a bigger story where the pieces should eventually fit together (like consistent naming, being able to create a single dashboard for the different metrics, etc.)

> cardinality shouldn't be an issue as long as it's not part of the counter name and only part of the attributes on the counter when incrementing it.

While I agree that the index name should be an attribute and not part of the metric name, from a memory and storage perspective, it doesn't really matter if the cardinality is encoded in an attribute vs the name.

> Do we want to track this for all data streams regardless of if they have failure stores enabled?

I think that would make sense but we should discuss this in some more detail. We should also align with the telemetry efforts @flash1293 is working on and think about who owns these metrics and which thresholds we want to set for ourselves.

> Failures that could have been stored but wont because failure store is disabled (?)

I'd expect those to be part of the `.rejected` metric. Maybe that metric could have an attribute to indicate whether the failure store would have caught it.

> * Documents sent to the failure store
>   * Suggestion `es.data_stream.ingest.documents.failed`

I think `failed` sounds a bit ambiguous and it's not intuitively clear what the difference to `rejected` is. Maybe `es.data_stream.ingest.documents.failure_store` would be a little more clear. I don't think it's important that all metrics end with a past tense verb. If you think otherwise, maybe `*.failure_stored`?

> * Documents rejected from the failure store (?)
>   * Suggestion `es.data_stream.ingest.documents.failure_store.rejected`

I'd also use the `es.data_stream.ingest.documents.rejected` metric for that. Potentially with a separate attribute that disambiguates regular rejections from those that the failure store rejected.

In general, it should be possible to get the failure rate by doing `es.data_stream.ingest.documents.rejected / es.data_stream.ingest.documents.total`. Therefore, all rejected documents should be included in `es.data_stream.ingest.documents.rejected`, no matter whether they could have been stored in the failure store or whether the failure store rejected them.

> Count of degraded documents (_ignored field)

Agree that this is out of scope for now. In the future we could just add another metric like `es.data_stream.ingest.documents.degraded`.

**Comment by felixbarny** [comment]
cc @elastic/es-storage-engine: this discussion is probably also relevant for you as it helps track our progress towards the goal of accepting all logs. It can help us identify areas where logs may still get dropped or added to the failure store unnecessarily so that we can prioritize efforts to minimize data loss or falling back on the failure store when there are other alternatives.

File: server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    public FailureStoreMetrics(MeterRegistry meterRegistry) {
+        this(
+            meterRegistry.registerLongCounter(METRIC_FAILURE_TOTAL, "ingest failure counter", "unit"),
+            meterRegistry.registerLongCounter(METRIC_REDIRECT_TOTAL, "ingest failure redirect counter", "unit")
+        );
+    }
+
+    public static void incrementForIndex(LongCounter counter, String index) {
+        counter.incrementBy(1, Map.of("index", index));
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:34]: Indices have an `index.mode` setting (which defaults to `standard`). I think the ability to "group by" index mode could be useful to distinguish different scenario and/or domains (logging, metrics and so on).

The `index.mode` setting is also a low cardinality value setting so I think adding such "dimension" to the attributes should be cheap. I see we are using the `index` instead which I expect to have much higher cardinality value that might result in explosion of time series to store (basically a counter incremented per each index).

Is there a specific reason why we need the "index" in the attributes?

**Comment by nielsbauman** [comment]
After several iterations of internal discussion, I think we reached a solution (proposal) that works for everyone. I just pushed a commit that should align the implementation with the discussed proposal. I don't expect the observability folks to look at the new changes themselves - as those are just implementation details of the proposal. I'll update the description of this PR as well to reflect the discussed proposal.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/IngestFailureStoreMetricsIT.java [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+        assertMeasurements(
+            measurements.get(FailureStoreMetrics.METRIC_FAILURE_STORE),
+            nrOfFailingDocs,
+            dataStream,
+            FailureStoreMetrics.ErrorLocation.PIPELINE
+        );
+        assertEquals(0, measurements.get(FailureStoreMetrics.METRIC_REJECTED).size());
+    }
+
+    public void testShardFailureNoFailureStore() throws IOException {
```
Comment (nielsbauman) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/IngestFailureStoreMetricsIT.java:140]: I wasn't able to come up with a way to make a redirect to the failure store fail on shard-level (i.e. get a rejection from the failure store). If anyone has any tips on that, feel free to share :)

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 7]
```diff
@@ -750,43 +748,70 @@ protected void doRun() {
...
+                                if (shouldKeep == false) {
+                                    String originalIndex = indexRequest.index();
+                                    // We only increment the total counter for dropped docs here, because these docs don't reach the code
+                                    // that ordinarily take care of that.
+                                    // We reuse `shouldStoreFailure` here to determine whether the index request targets a data stream,
+                                    // because we only want to track these metrics for data streams.
+                                    Optional<Boolean> storeFailure = shouldStoreFailure.apply(originalIndex);
+                                    if (storeFailure.isPresent()) {
+                                        failureStoreMetrics.incrementTotal(originalIndex);
                                     }
```
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java:764]: Reusing `shouldStoreFailure` here isn't ideal, but it avoids having to duplicate some logic. Maybe a method rename would already be sufficient?

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 8]
```diff
@@ -216,6 +225,25 @@ protected void doInternalExecute(
         );
     }
 
+    /**
+     * Track the number of index requests in our APM metrics. We'll track almost all docs here (pipeline or no pipeline,
+     * failure store or original), but some docs don't reach this place (dropped and rejected docs), so we increment for those docs in
+     * different places.
```
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:231]: Ideally, we'd increment the `total` counter in one place, but I couldn't come up with a way to make that happen (because of the reroute vs. dropping "contradiction").

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 9]
```diff
@@ -750,43 +748,70 @@ protected void doRun() {
...
-                                @Override
-                                public void onFailure(Exception e) {
+                            @Override
+                            public void onFailure(Exception e) {
+                                final String originalIndex = indexRequest.index();
+                                String errorType = ElasticsearchException.getExceptionName(ExceptionsHelper.unwrapCause(e));
+                                // If `storeFailure` is true, we store the failure. If it's false, the target is a data stream,
+                                // but it doesn't have the failure store enabled. If it's empty, the target wasn't a data stream.
+                                Optional<Boolean> storeFailure = shouldStoreFailure.apply(originalIndex);
+                                if (storeFailure.orElse(false)) {
```
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java:779]: Not really a fan of this `Optional<Boolean>` myself tbh. My struggle was to avoid having to check a second time whether the request targets a data stream when `storeFailure` is `false`.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 10]
```diff
@@ -750,43 +748,70 @@ protected void doRun() {
                         final org.elasticsearch.script.Metadata originalDocumentMetadata = ingestDocument.getMetadata().clone();
                         // the document listener gives us three-way logic: a document can fail processing (1), or it can
                         // be successfully processed. a successfully processed document can be kept (2) or dropped (3).
-                        final ActionListener<IngestPipelinesExecutionResult> documentListener = ActionListener.runAfter(
-                            new ActionListener<>() {
-                                @Override
-                                public void onResponse(IngestPipelinesExecutionResult result) {
-                                    assert result != null;
```
Comment (nielsbauman) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java:757]: This whole refactor of `documentListener` and `exceptionHandler` isn't really necessary anymore (it's a remnant of one of my previous approaches where it was necessary), but I think it looks a bit cleaner so I decided to leave it in. If people feel differently, I'm happy to revert the changes.

File: server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        failureStoreCounter = meterRegistry.registerLongCounter(
+            METRIC_FAILURE_STORE,
+            "number of documents that got redirected to the failure store",
+            "unit"
+        );
+        rejectedCounter = meterRegistry.registerLongCounter(METRIC_REJECTED, "number of documents that were rejected", "unit");
+    }
+
+    /**
+     * This counter tracks the number of documents that we <i>tried</i> to index into a data stream. This includes dropped documents.
```
Comment (gmarouli) [inline server/src/main/java/org/elasticsearch/action/bulk/FailureStoreMetrics.java:46]: I would suggest to rephrase the last sentence to something more specific. If I am unfamiliar with the drop processor I would not have understood what you meant.
`This includes dropped documents.` -> `This includes documents that were "dropped" by a pipeline.`.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 7 above)
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:764]: I see your concern, how about splitting it into two predicates or something and use the appropriate one each time?

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 9 above)
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:779]: Is it that big of an issue?

**Review by gmarouli — CHANGES_REQUESTED** [review]
Hey @nielsbauman really excited to see this moving! It looks good, here are a few small comments.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 9 above)
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:779]: Hmm well we'd be doing it twice for every index request that doesn't have the failure store enabled. I'm usually inclined to pay extra care to doing things for every index request, but that's not based on any experimentation on the impact of doing a second lookup.

Alternatively, I could change `shouldStoreFailure` to a function that returns a failure store _candidate_ just like I do in `BulkOperation`. I'm also trying to think about the future where the decision to redirect a failure or not might be influenced by the request itself.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/IngestFailureStoreMetricsIT.java [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+        dataStream = "data-stream-" + randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+        pipeline = "pipeline-" + randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+        logger.info(
+            "--> running [{}] with data stream [{}], template [{}] and pipeline [{}]",
+            getTestName(),
+            dataStream,
+            template,
+            pipeline
+        );
+    }
```
[the remark this answers is not in the document]
Comment (nielsbauman) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/IngestFailureStoreMetricsIT.java:77]: Ah sure. I don't really have an issue with this pattern because I've seen it in a bunch of places and thus automatically associate it with only generating names. But I'm completely fine with your suggestions :+1:.

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/IngestFailureStoreMetricsIT.java (hunk 6 above)
Comment (nielsbauman) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/IngestFailureStoreMetricsIT.java:140]: Perfect suggestion! Works like a charm :smile:

**Review by jbaiera — CHANGES_REQUESTED** [review]

**Comment by jbaiera** [comment]
Hit "request changes" before actually typing my review text 🙄 

I think the changes are looking good, I mostly have some concerns around the refactoring in the ingest service. I wonder if we could increase the test coverage to include situations like updating the index field and using the reroute processor

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 13]
```diff
@@ -750,43 +748,72 @@ protected void doRun() {
...
+                                assert shouldKeep != null;
+                                if (shouldKeep == false) {
+                                    String originalIndex = indexRequest.index();
+                                    // We only increment the total counter for dropped docs here, because these docs don't reach the code
+                                    // that ordinarily take care of that.
+                                    // We reuse `resolveFailureStore` here to determine whether the index request targets a data stream,
+                                    // because we only want to track these metrics for data streams.
+                                    Boolean failureStoreResolution = resolveFailureStore.apply(originalIndex);
+                                    if (failureStoreResolution != null) {
+                                        failureStoreMetrics.incrementTotal(originalIndex);
```
[the remark this answers is not in the document]
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:763]: Ah good point, I think it will associate the write to the alias. Let me see if I can come up with a good way to avoid that.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java [hunk 14]
```diff
@@ -888,14 +914,7 @@ private void executePipelines(
...
         ingestDocument.resetReroute();
-        final String originalIndex = indexRequest.indices()[0];
-        final Consumer<Exception> exceptionHandler = (Exception e) -> {
-            if (shouldStoreFailure.test(originalIndex)) {
-                listener.onResponse(IngestPipelinesExecutionResult.failAndStoreFor(originalIndex, e));
-            } else {
-                listener.onFailure(e);
-            }
-        };
+        final String originalIndex = indexRequest.index();
```
[the remark this answers is not in the document]
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:917]: https://github.com/elastic/elasticsearch/pull/108279#discussion_r1681141168

I added `IngestFailureStoreMetricsIT.java#testRerouteSuccessfulCorrectName` to verify that rerouting works correctly. Or did you have something else in mind?

[the remark this answers is not in the document]
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:936]: Fixed in [fd644eb](https://github.com/elastic/elasticsearch/pull/108279/commits/fd644eb5e29bf3b4aacd80effe1a4f13ce28b8d3)

[the remark this answers is not in the document]
Comment (jbaiera) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:936]: There's a few scenarios where I think this refactoring could still hit snags. Let's assume we have a pipeline that uses the set processor to update the `_index` field on the document to a different data stream. 

If the document later fails processing inside that pipeline it looks like the failure will be attributed to the original pipeline still since the metadata isn't updated on the document.

However, the metadata is eventually updated on the original document request. We then carry out a few more checks:

1. A document that has a self reference would fail ingestion because it is not serializable.
2. A document that has been passed to a pipeline that it has already visited during ingestion will fail to avoid infinite looping.
3. A document that has its index name changed by its final pipeline will fail because this violates an invariant of final pipelines.

In all these cases, it looks like the failure will be attributed to the new data stream that was given in the set pipeline, even though the error in the processing did not occur there.

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: Fixed in [fd644eb](https://github.com/elastic/elasticsearch/pull/108279/commits/fd644eb5e29bf3b4aacd80effe1a4f13ce28b8d3)

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:936]: Sorry my comment `Fixed in [fd644eb](https://github.com/elastic/elasticsearch/pull/108279/commits/fd644eb5e29bf3b4aacd80effe1a4f13ce28b8d3)` was intended for our other discussion.

About your points here, I was under the impression that we wanted to use the destination data stream in/after reroutes? With the reasoning that if a user reroutes a document to a different data stream (and it failds), the Observability team doesn't want to be paged on it.

[the remark this answers is not in the document]
Comment (jbaiera) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:936]: We do indeed! But the reroute operation isn't actually completed until the pipeline execution it was called from is complete. That execution is only completed after those above validations occur. If those validations fail, we want to attribute the failure to the data stream we were writing to when the pipeline execution began. This is why the old code had an error handler that captured the original index within it, for a stable handle to that info.

[the remark this answers is not in the document]
Comment (jbaiera) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java:936]: > if a user reroutes a document to a different data stream (and it failds), the Observability team doesn't want to be paged on it.

I'm not familiar with any plans to do paging in this way. That said, I don't think that's a reasonable approach to building the alert: If the user does something that breaks processing in a custom pipeline and doesn't reroute the document it still will have a failure showing up in the original data stream. 

If we do move forward with building those kinds of alerts, I would rather us attribute the document to the correct data stream and attempt to limit alerting based on the pipeline name that the failure occurs in.

[the remark this answers is not in the document]
Comment (nielsbauman) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: Done in [6a493c5](https://github.com/elastic/elasticsearch/pull/108279/commits/6a493c50b15bbf73ae394a64334133c187b62544)

**Review by jbaiera — APPROVED** [review]
LGTM! Thanks for all the work on this one!

**Review by gmarouli — APPROVED** [review]
LGTM, what a ride! Thank you @nielsbauman !
